### ISSUE TITLE
DBS3Upload fails to insert block with "413 Request Entity Too Large"

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11316]
**Impact of the bug**
WMAgent

**Describe the bug**
While investigating lack of central services and wmagent information in WMStats "agent info" tab, I actually spotted a new error message in the DBS3Upload component of vocms0252 (other agents are still to be checked), which fails to upload block(s) with this error:
```
413 Request Entity Too Large
```

After adding one debug line to the code, I actually see that we have almost 500MB of data to post to DBSServer [1], which might be directly related to the slowness in CMSWEB services that Todor has been mentioning for a couple of days now.

**How to reproduce it**
Insert a very large (~480MB) block into DBS.

**Expected behavior**
Of course we should make sure blocks are much smaller than this, but this is a development that can't be made and deployed within the next few days.

So, we should investigate who is yielding this HTTP error:
1) Apache frontends?
2) Nginx?
3) DBS2Go server?

and evaluate whether this limit can be raised, such that we can work on this limits in the agent in a non-panic mode.

Longer term development can be tracked in: https://github.com/dmwm/WMCore/issues/10264

**Additional context and error message**
[1]
```
2022-10-05 04:53:48,415:140243882792768:INFO:DBSUploadPoller:About to call insert block for: /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau0p01mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#3b2d7278-09a9-49a9-8c7b-4b6530b389ed with a payload size of 485.5212106704712 MB
2022-10-05 04:53:59,748:140243882792768:ERROR:DBSUploadPoller:Error trying to process block /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau0p01mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#3b2d7278-09a9-49a9-8c7b-4b6530b389ed through DBS. Error: <html>
<head><title>413 Request Entity Too Large</title></head>
<body>
<center><h1>413 Request Entity Too Large</h1></center>
<hr><center>nginx</center>
</body>
</html>
Traceback (most recent call last):
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-client/4.0.10/lib/python3.8/site-packages/dbs/apis/dbsClient.py", line 461, in __callServer
    self.http_response = method_func(self.url, method, params, data, request_headers)
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-pycurl/3.17.7-comp2/lib/python3.8/site-packages/RestClient/RestApi.py", line 42, in post
    return http_request(self._curl)
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-pycurl/3.17.7-comp2/lib/python3.8/site-packages/RestClient/RequestHandling/HTTPRequest.py", line 62, in __call__
    raise HTTPError(effective_url, http_code, http_response.msg, http_response.raw_header, http_response.body)
RestClient.ErrorHandling.RestClientExceptions.HTTPError: HTTP Error 413: Request Entity Too Large

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/wmagentpy3/2.1.2/lib/python3.8/site-packages/WMComponent/DBS3Buffer/DBSUploadPoller.py", line 93, in uploadWorker
    dbsApi.insertBulkBlock(blockDump=block)
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-client/4.0.10/lib/python3.8/site-packages/dbs/apis/dbsClient.py", line 630, in insertBulkBlock
    result =  self.__callServer("bulkblocks", data=blockDump, callmethod='POST' )
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-client/4.0.10/lib/python3.8/site-packages/dbs/apis/dbsClient.py", line 473, in __callServer
    self.__parseForException(data)
  File "/data/srv/wmagent/v2.1.2/sw/slc7_amd64_gcc630/cms/py3-dbs3-client/4.0.10/lib/python3.8/site-packages/dbs/apis/dbsClient.py", line 499, in __parseForException
    raise http_error
RestClient.ErrorHandling.RestClientExceptions.HTTPError: HTTP Error 413: 
URL=https://cmsweb-prod.cern.ch:8443/dbs/prod/global/DBSWriter/bulkblocks
Code=413
Message=Request Entity Too Large
Header=HTTP/1.1 100 Continue
```

**Comment by todor-ivanov** [comment]
Thanks @amaltaro. As mentioned in the communication with CMSWEB Team they have implemented the so proposed new limits in the frontend, but we still see the error fro two blocks at least:
```2022-10-05 08:22:41,578:139814743262976:INFO:DBSUploadPoller:About to call insert block for: /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau0p01mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7
-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#3b2d7278-09a9-49a9-8c7b-4b6530b389ed with a payload size of 485.5212106704712 MB
2022-10-05 08:22:53,382:139814743262976:ERROR:DBSUploadPoller:Error trying to process block /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau0p01mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-
Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#3b2d7278-09a9-49a9-8c7b-4b6530b389ed through DBS. Error: <html>
<head><title>413 Request Entity Too Large</title></head>
```


And this also does not seem to be the only reason issue we are having regarding central services. I fail to relate this issue as a direct cause for the missing central services components in WMStats. It could still be sign of something (some load) the system finds it difficult to digest, but this is only on the agent's  side at least. I am creating a separate issue for this investigation.

**Comment by todor-ivanov** [comment]
Well, even before I manage to create a separate issue, all the symptoms for the central services components (and wmstats in general) just disappeared... 

The error at the agent still persists. 
FYI @amaltaro

**Comment by vkuznet** [comment]
@amaltaro , the DBS codebase does not have any call to issue 413 HTTP error. Most likely it is our nginx. @muhammadimranfarooqi  could you please check nginx configuration settings to see which is upper limit it has on payload size, I have vague memory that it is exactly 500MB.

**Comment by amaltaro** [comment]
Looking at the logs, the only strange record I can find is:
```
2022/10/05 13:56:16 [warn] 41#41: *1597788 a client request body is buffered to a temporary file /tmp/client-body/0000000548, client: xxx.xxx.xxx.xxx, server: cmsweb-k8s-prodsrv.cern.ch, request: "POST /dbs/prod/global/DBSWriter/bulkblocks HTTP/1.1", host: "cmsweb-k8s-prodsrv.cern.ch"
```

does anyone understand it?

**Comment by amaltaro** [comment]
You can find it in a few pods, but here is one of them `cern-magnum-ingress-nginx-controller-96qkn`.

**Comment by vkuznet** [comment]
and message is quite clear, the request is buffered to temp file, meaning that request is big and should be uploaded in chunks. It seems to me that nginx split the incoming body into small parts in order to uploaded upstream, that's it.

**Comment by vkuznet** [comment]
once, again I think it is configuration option in nginx which we already increased once. And, once again my advise to not increase it but rather stop producing this big payloads.

**Comment by muhammadimranfarooqi** [comment]
> @amaltaro , the DBS codebase does not have any call to issue 413 HTTP error. Most likely it is our nginx. @muhammadimranfarooqi could you please check nginx configuration settings to see which is upper limit it has on payload size, I have vague memory that it is exactly 500MB. If I'm not mistaken at some point one of the group (may be this one) requested to increase a limit from 300MB to 500MB and we did that. I always advised to avoid such increase since the argument of "we need it to bypass current limitation" leads to new limitations when we hit them. And, this seems to be the case here. We need to fix root of the problem to not make such payloads.

I increased it today to 600m as per Alan's request

**Comment by amaltaro** [comment]
> I think it is configuration option in nginx which we already increase it once. And, once again my advise to not increase it but rather stop producing this big payloads.

Yes, we know. Pull requests are welcomed: https://github.com/dmwm/WMCore/issues/10264

**Comment by vkuznet** [comment]
The remedy here is not to increase threshold on nginx but fix the code to use gzip encoding when you pass your payload. This will allow to use large blocks and lumis which is separate issue from payload. Therefore what is required here is the following:
- add gzip Encoding in DBS3Upload
- gzip payload
- send gzip'ed  payload to DBS server
- reverse nginx contrain to 200MB (if I recall that was our initial limit)
- address #10264 separately since it is data management decision about size of dataset/blocks we produce but it has nothing to do with HTTP requests we use and how we send the large JSON data to our upstream server

@amaltaro , please evaluate this proposal and if you agree I can provide gzip encoding to DBS3Upload (if you'll point me to concrete place where this payload is created).

**Comment by amaltaro** [comment]
Thanks for reminding me of this feature, Valentin. You actually implemented this feature as of a few months ago, so all that I needed to do in vocms0252 was to switch the following agent configuration and restart DBS3Upload:
```
config.DBS3Upload.gzipEncoding = True  # False
```

With the payload compression enabled, one of the blocks did go through and the http request took around 30 seconds to be fulfilled (note that json.dumps() report a total size of 444MB of data):
```
2022-10-06 04:54:38,375:139857554515776:INFO:DBSUploadPoller:About to call insert block for: /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau10p0mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#a2b77cda-eb9c-4f45-867b-1052bed8d7a0 with a payload size of 444.49663066864014 MB
```

however, a block with a slightly larger payload failed with a proxy error (timeout in the frontend):
```
2022-10-06 04:55:23,036:139857554515776:INFO:DBSUploadPoller:About to call insert block for: /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau0p01mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#3b2d7278-09a9-49a9-8c7b-4b6530b389ed with a payload size of 485.5212106704712 MB
2022-10-06 05:00:04,432:139857554515776:ERROR:DBSUploadPoller:Error trying to process block /BcToNMuX_NToEMuPi_SoftQCD_b_mN5p5_ctau10p0mm_TuneCP5_13TeV-pythia8-evtgen/RunIIAutumn18NanoAODv7-Custom_RDStar_BParking_Nano02Apr2020_102X_upgrade2018_realistic_v21-v1/NANOAODSIM#a2b77cda-eb9c-4f45-867b-1052bed8d7a0 through DBS. Error: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>502 Proxy Error</title>.
...
```

Let's see if this one will go through in the next cycle.

**Comment by vkuznet** [comment]
@amaltaro , after #11325 is applied please confirm explicitly if you agree to revert back the nginx threshold, if not please provide your arguments. I would like to put it back to 200MB level specifically to avoid bad practices from developers to send very large data payloads to CMSWEB.

**Comment by amaltaro** [comment]
I don't think having such limits at Nginx is our best model here. The reason is that it's hard to debug, log files are not accessible - unless you have access to that namespace - and error messages are just generic.
Given that it's a component in the middle of frontends and backends, it makes things even worse to debug.

So, I am in favor of actually removing these limits, or at least keep the limit that we had before (not 200MB).

**Comment by vkuznet** [comment]
@amaltaro I think it is the opposite of what you said. Since we identified that large payload leads to 413 Request Entity Too Large HTTP error. This is a very clear error of the underlying issue, i.e. once someone hit this error we know that such client send too much data. That's it! While if we remove the limit then we'll be in trouble debugging why we may see weird behavior. For instance, I can imaging that someone will upload 1GB payload and it will affect performance of our FE, i.e. it will affect other services. This will be *VERY DIFFICULT* to debug indeed. We will keep the limit.

**Comment by amaltaro** [comment]
If you want to keep the limits, then please make sure logs and documentation on how to access and debug Nginx issues is available.
This is a new'ish "feature" of the k8s environment and it has never been discussed with service responsibles. If you want to give projects full control of their service, then we must be able to see all the pieces and logs in the chain, without sending 10s of messages and bugging people to help with that.

**Comment by vkuznet** [comment]
Yes, it is fair request and I created ticket for that https://its.cern.ch/jira/browse/CMSKUBERNETES-189, but it is not new'ish feature. The nginx is part of k8s by default and it has default limits. We **discover** them as we start using more and more k8s infrastructure, but it is not something we intentionally introduced to make life of dev-teams harder.
